iceoryx_binding_c/source/binding_c_publisher.cpp

```diff
@@ -0,0 +1,95 @@
+// Copyright (c) 2020 by Robert Bosch GmbH. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "iceoryx_posh/internal/popo/ports/publisher_port_user.hpp"
+
+using namespace iox;
+using namespace iox::cxx;
+using namespace iox::popo;
+using namespace iox::capro;
+using namespace iox::mepoo;
+
+extern "C" {
+#include "iceoryx_binding_c/publisher.h"
+}
+
+PublisherPortData*
+Publisher_new(const char* service, const char* instance, const char* event, const uint64_t historyCapacity)
+{
+    return new PublisherPortData(
+        ServiceDescription{
+            IdString(TruncateToCapacity, service),
+            IdString(TruncateToCapacity, instance),
+            IdString(TruncateToCapacity, event),
+        },
+        "JoinTheChurchOfHypnotoad!",
+        nullptr,
+        historyCapacity);
+}
+
+void Publisher_delete(PublisherPortData* const self)
+{
+    delete self;
+}
+
+iox_popo_AllocationError
+Publisher_allocateChunk(PublisherPortData* const self, void** const chunk, const uint32_t payloadSize)
+{
+    auto result =
+        PublisherPortUser(self).allocateChunk(payloadSize).and_then([&](ChunkHeader* h) { *chunk = h->payload(); });
+    if (result.has_error())
+    {
+        return static_cast<iox_popo_AllocationError>(static_cast<int>(result.get_error()));
```

Comment:
Do we have to rely on the same order in two diffferent enums? How can this be made save?

Reply:
not really nice, we either have to give up type safety in our C++ codebase by using enums instead of enum classes or create a second enum. Personally, I would also opt for the approach with the second enum. We could check if the enums match to make this save, e.g.
```
enum CFoo {
    BAR,
    BAZ
};
enum class CppFoo {
    BAR,
    BAZ
};

// somewhere in the tests
constexpr int NUMBER_OF_ENUM_VALUES {2};
for(int i = 0; i < NUMBER_OF_ENUM_VALUES; ++i) {
    CppFoo foo = static_cast<CppFoo>(i);
    switch (foo) {
        case CppFoo::BAR:
            EXPECT_THAT(static_cast<int>(foo), BAR);
            break;
        case CppFoo::BAZ:
            EXPECT_THAT(static_cast<int>(foo), BAZ);
            break;
        default:
            constexpr bool ENUM_MISSMATCH {true};
            EXPECT_FALSE(ENUM_MISSMATCH);
    }
}
```
We still have to keep in mind to update the test or maybe add something like `LAST_VALUE_PLEASE_ADD_ITEMS_BEFORE_THIS` to the enum to always get the right number of items in the enum

Reply:
This is already contained in the tests. But I am using a switch case statement to ensure the correct translation!